IPC::Connection cannot be bound to arbitrary run loop

mattwoodrow opened the pull request:
#### e23c41101c83695756b09b8741c7529e7b007212
<pre>
IPC::Connection cannot be bound to arbitrary run loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=246336">https://bugs.webkit.org/show_bug.cgi?id=246336</a>
rdar://problem/101030022

Reviewed by Chris Dumez.

Make it possible to bind IPC::Connection::Client to a particular
run loop via Connection::open(Client&amp;, RunLoop&amp;). The messages
will be dispatched to the passed run loop.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::runLoop):
(IPC::Connection::Connection):
(IPC::Connection::~Connection):
(IPC::Connection::connection):
(IPC::Connection::open):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::sendSyncMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::dispatchDidCloseAndInvalidate):
(IPC::Connection::enqueueIncomingMessage):
(IPC::Connection::dispatchMessage):
(IPC::Connection::dispatchSyncStateMessages):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::wakeUpRunLoop):
(IPC::Connection::dispatchClient):
(IPC::Connection::postConnectionDidCloseOnConnectionWorkQueue): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/255814@main">https://commits.webkit.org/255814@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b92e12c30195aac7a03300df89eb087408f050a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93657 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2852 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103308 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163631 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97653 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2863 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31130 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86009 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99319 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80101 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83963 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37515 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35357 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4014 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39233 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41168 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -336,12 +335,10 @@ Connection::~Connection()
-// It is safe because this function is only called on the main thread and Connection objects are only
-// constructed / destroyed on the main thread.
-Connection* Connection::connection(UniqueID uniqueID) WTF_IGNORES_THREAD_SAFETY_ANALYSIS
+Connection* Connection::connection(UniqueID uniqueID)
```
Shouldn't this return a `RefPtr<Connection>`? I don't see how it is ever OK to return a raw pointer if the Connection object can get destroyed on another thread now. Locking *inside* the function doesn't help if we don't protect the pointer we're returning.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -475,7 +472,6 @@ bool Connection::platformPrepareForOpen()
 
 void Connection::invalidate()
 {
-    ASSERT(RunLoop::isMain());
```
Could we maybe keep all these assertions and just update them to make sure that we're on the expected run loop now? Dropping assertions always makes me worried.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -575,7 +571,9 @@ Timeout Connection::timeoutRespectingIgnoreTimeoutsForTesting(Timeout timeout) c
-    ASSERT(RunLoop::isMain());
+    if (!isValid())
+        return nullptr;
+    assertIsCurrent(m_syncState->runLoop());
```
I guess this is a suitable replacement to the `ASSERT(RunLoop::isMain());` above?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1074,10 +1056,22 @@ void Connection::dispatchSyncMessage(Decoder& decoder)
+
+void Connection::dispatchDidCloseAndInvalidate()
+{
+    dispatchClient([protectedThis = Ref { *this }] {
```
Can we please call this `dispatchToClient`? `dispatchClient` is confusing IMO, since it is not the client we're dispatching.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1124,17 +1118,21 @@ void Connection::enqueueIncomingMessage(std::unique_ptr<Decoder> incomingMessage
 
 void Connection::dispatchMessage(Decoder& decoder)
 {
-    ASSERT(RunLoop::isMain());
```
Could this assertion be updated instead of removed?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1261,14 +1258,14 @@ void Connection::dispatchOneIncomingMessage()
 
 void Connection::dispatchSyncStateMessages()
 {
-    ASSERT(RunLoop::isMain());
```
Could this be updated instead of removed?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1293,7 +1290,7 @@ void Connection::dispatchIncomingMessages()
         // spins a nested run loop while we're dispatching a message. Note that this means we can re-enter this method.
         if (!m_incomingMessages.isEmpty()) {
-            RunLoop::main().dispatch([protectedThis = Ref { *this }] {
+            m_syncState->runLoop().dispatch([protectedThis = Ref { *this }] {
```
Seems confusing? Why are we getting a runloop from m_syncState when this code doesn't appear to be about sync messages at all?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1356,7 +1353,18 @@ CompletionHandler<void(Decoder*)> takeAsyncReplyHandler(Connection& connection,
+}
+
+template<typename F>
+void Connection::dispatchClient(F&& clientRunLoopTask)
```
Confusing name.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1356,7 +1353,18 @@ CompletionHandler<void(Decoder*)> takeAsyncReplyHandler(Connection& connection,
-    RunLoop::main().wakeUp();
+    if (!isValid())
+        return;
+    m_syncState->runLoop().wakeUp();
```
Confusing to be getting a runLoop from m_syncState when this code is not about sync IPC

mattwoodrow replied:
I moved these into a helper that explains it a bit better, and asserts the required conditions.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1356,7 +1353,18 @@ CompletionHandler<void(Decoder*)> takeAsyncReplyHandler(Connection& connection,
+    Locker lock { m_incomingMessagesLock };
+    if (!m_syncState)
+        return;
+    m_syncState->runLoop().dispatch(WTFMove(clientRunLoopTask));
```
Confusing to be getting a runLoop from m_syncState when this code is not about sync IPC

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -613,8 +618,6 @@ void moveTuple(std::tuple<A...>&& a, std::tuple<B...>& b)
 template<typename T> Connection::SendSyncResult<T> Connection::sendSync(T&& message, uint64_t destinationID, Timeout timeout, OptionSet<SendSyncOption> sendSyncOptions)
 {
     static_assert(T::isSync, "Sync message expected");
-    RELEASE_ASSERT(RunLoop::isMain());
```
Can we update the assertion instead of dropping it?

mattwoodrow replied:
These assertions in the header all come from functions that call directly into something else which also has the assert.

The assert now requires the `isValid()` check too, so having it in both is a bit painful (but I can add them back if you'd prefer).

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -641,7 +644,6 @@ template<typename T> Connection::SendSyncResult<T> Connection::sendSync(T&& mess
 
 template<typename T> bool Connection::waitForAndDispatchImmediately(uint64_t destinationID, Timeout timeout, OptionSet<WaitForOption> waitForOptions)
 {
-    RELEASE_ASSERT(RunLoop::isMain());
```
Can we update the assertion instead of dropping it?

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -654,7 +656,6 @@ template<typename T> bool Connection::waitForAndDispatchImmediately(uint64_t des
 
 template<typename T> bool Connection::waitForAsyncCallbackAndDispatchImmediately(uint64_t destinationID, Timeout timeout)
 {
-    RELEASE_ASSERT(RunLoop::isMain());
```
Can we update the assertion instead of dropping it?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1245,6 +1244,21 @@ size_t Connection::numberOfMessagesToProcess(size_t totalMessages)
+        assertIsCurrent(m_syncState->runLoop());
+
+    // Our syncState is specific to the RunLoop we have been
+    // bound to during open(), so we can retreive the RunLoop
```
typo: retreive

cdumez reviewed (approved):
LGTM

kkinnunen-apple committed:
IPC::Connection cannot be bound to arbitrary run loop
https://bugs.webkit.org/show_bug.cgi?id=246336
rdar://problem/101030022

Reviewed by Chris Dumez.

Make it possible to bind IPC::Connection::Client to a particular
run loop via Connection::open(Client&, RunLoop&). The messages
will be dispatched to the passed run loop.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::runLoop):
(IPC::Connection::Connection):
(IPC::Connection::~Connection):
(IPC::Connection::connection):
(IPC::Connection::open):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::sendSyncMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::dispatchDidCloseAndInvalidate):
(IPC::Connection::enqueueIncomingMessage):
(IPC::Connection::dispatchMessage):
(IPC::Connection::dispatchSyncStateMessages):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::wakeUpRunLoop):
(IPC::Connection::dispatchClient):
(IPC::Connection::postConnectionDidCloseOnConnectionWorkQueue): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/255814@main
